ferveo/src/refresh.rs

```diff
@@ -0,0 +1,384 @@
+/// From PSS paper, section 4.2.3, (https://link.springer.com/content/pdf/10.1007/3-540-44750-4_27.pdf)
+pub fn apply_updates_to_private_share<E: Pairing>(
+    private_key_share: &PrivateKeyShare<E>,
+    share_updates: &[E::G2],
+) -> PrivateKeyShare<E> {
+    let private_key_share = share_updates
+        .iter()
+        .fold(
+            private_key_share.private_key_share.into_group(),
+            |acc, delta| acc + delta,
+        )
+        .into_affine();
+    PrivateKeyShare { private_key_share }
+}
+
+/// From the PSS paper, section 4.2.4, (https://link.springer.com/content/pdf/10.1007/3-540-44750-4_27.pdf)
+pub fn recover_share_from_updated_private_shares<E: Pairing>(
+    x_r: &E::ScalarField,
+    domain_points: &[E::ScalarField],
+    updated_private_shares: &[PrivateKeyShare<E>],
+) -> PrivateKeyShare<E> {
+    // Interpolate new shares to recover y_r
+    let lagrange = lagrange_basis_at::<E>(domain_points, x_r);
+    let prods = zip_eq(updated_private_shares, lagrange)
+        .map(|(y_j, l)| y_j.private_key_share.mul(l));
+    let y_r = prods.fold(E::G2::zero(), |acc, y_j| acc + y_j);
+
+    PrivateKeyShare {
+        private_key_share: y_r.into_affine(),
+    }
+}
+
+// SHARE REFRESH FUNCTIONS:
+
+pub fn prepare_share_updates_for_refresh<E: Pairing>(
+    domain_points: &[E::ScalarField],
+    h: &E::G2Affine,
+    threshold: usize,
+    rng: &mut impl RngCore,
+) -> Vec<E::G2> {
+    // Update polynomial has root at 0
+    prepare_share_updates_with_root::<E>(
+        domain_points,
+        h,
+        &E::ScalarField::zero(),
+        threshold,
+        rng,
+    )
+}
+
+// UTILS:
+
+fn prepare_share_updates_with_root<E: Pairing>(
+    domain_points: &[E::ScalarField],
+    h: &E::G2Affine,
+    root: &E::ScalarField,
+    threshold: usize,
+    rng: &mut impl RngCore,
+) -> Vec<E::G2> {
+    // Generate a new random polynomial with defined root
+    let d_i = make_random_polynomial_with_root::<E>(threshold - 1, root, rng);
+
+    // Now, we need to evaluate the polynomial at each of participants' indices
+    domain_points
+        .iter()
+        .map(|x_i| {
+            let eval = d_i.evaluate(x_i);
+            h.mul(eval)
+        })
+        .collect()
+}
+
+pub fn make_random_polynomial_with_root<E: Pairing>(
+    degree: usize,
+    root: &E::ScalarField,
+    rng: &mut impl RngCore,
+) -> DensePolynomial<E::ScalarField> {
+    // [c_0, c_1, ..., c_{degree}] (Random polynomial)
+    let mut poly = DensePolynomial::<E::ScalarField>::rand(degree, rng);
+
+    // [0, c_1, ... , c_{degree}]  (We zeroize the free term)
+    poly[0] = E::ScalarField::zero();
+
+    // Now, we calculate a new free term so that `poly(root) = 0`
+    let new_c_0 = E::ScalarField::zero() - poly.evaluate(root);
+    poly[0] = new_c_0;
+
+    // Evaluating the polynomial at the root should result in 0
+    debug_assert!(poly.evaluate(root) == E::ScalarField::zero());
+    debug_assert!(poly.coeffs.len() == degree + 1);
```

Comment:
Just a comment on this macro usage:
`debug_assert!` doesn't show up in a `release` target, i.e. this assertion would be removed in production. If we want this assertion to actually throw in production, we should use `assert!` instead. I didn't think it would be necessary, so I opted for the former.